overlord/hookstate/hookmgr.go

```diff
@@ -279,47 +280,66 @@ func (m *HookManager) undoRunHook(task *state.Task, tomb *tomb.Tomb) error {
 		return fmt.Errorf("cannot extract undo hook setup from task: %s", err)
 	}
 
-	return m.runHook(task, tomb, snapst, hooksup)
+	return m.runHookForTask(task, tomb, snapst, hooksup)
 }
 
-func (m *HookManager) runHook(task *state.Task, tomb *tomb.Tomb, snapst *snapstate.SnapState, hooksup *HookSetup) error {
+func (m *HookManager) EphemeralRunHook(ctx context.Context, hooksup *HookSetup, contextData map[string]interface{}) (*Context, error) {
+	tomb, _ := tomb.WithContext(ctx)
+
+	var snapst snapstate.SnapState
+	m.state.Lock()
+	err := snapstate.Get(m.state, hooksup.Snap, &snapst)
+	m.state.Unlock()
+	if err != nil {
+		return nil, fmt.Errorf("cannot run ephemeral hook %q for snap %q: %v", hooksup.Hook, hooksup.Snap, err)
+	}
+
+	return m.runHookCommon(nil, tomb, &snapst, hooksup, contextData)
+}
+
+func (m *HookManager) runHookForTask(task *state.Task, tomb *tomb.Tomb, snapst *snapstate.SnapState, hooksup *HookSetup) error {
+	_, err := m.runHookCommon(task, tomb, snapst, hooksup, nil)
+	return err
+}
+
+func (m *HookManager) runHookCommon(task *state.Task, tomb *tomb.Tomb, snapst *snapstate.SnapState, hooksup *HookSetup, contextData map[string]interface{}) (*Context, error) {
```

Comment:
I have an idea, not a blocker, to clean this up a bit further,  if we added to Context Logf and Errorf that redirect to the task or use the general logger otherwise maybe?  then we create the context beforehand (it doesn't have side effects anyway) and pass it in to this called runHookInContext,  the params would change a bit

Reply:
I pushed this cleanup as https://github.com/snapcore/snapd/compare/master...mvo5:hookstate-run2?expand=1 and I can cleanup (need some tests) and propose once this is merged (and if it looks like a reasonable win :)

Reply:
the cleanup looks good but two remarks:

* probably we shouldn't lock only one one path of Logf Errorf but instead require that the callers uses Context.Lock/Unlock. This means Logf and Errorf needs to be added to the doc comment of Lock (anywhere else?)
* runHookInContext doesn't need to return the context anymore

we probably now can also merge  NewContext and initEphemeralContextData in some kind of newEphemeralContext dedicated to that use case